### PR TITLE
Guardrails: Expose Dedicated Endpoint

### LINKED ISSUE #935: Guardrails: Expose dedicated endpoint
**Is your feature request related to a problem?**
We currently do not have a dedicated endpoint for using guardrails directly, which can complicate the integration process and delay development.

**Describe the solution you'd like**
- Expose an endpoint that allows users to access only guardrails.
- Ensure that this endpoint does not route through the LLM/call endpoint.

<details><summary>Original issue</summary>

TBD.

Requirements will be added when things are clear

</details>

### LINKED ISSUE #1033: Guardrails: Standalone access endpoints
**Is your feature request related to a problem?**
We need to access standalone guardrails without making a call to the main LLM. This is essential for our current requirement from TAP.

**Describe the solution you'd like**
- Build endpoints to support this requirement.

<details><summary>Original issue</summary>

- Build endpoints to support the requirement
- Current requirement from TAP

**Comments:** iterations pending

</details>

### PR DESCRIPTION
**Pull request by Prajna1999**
## Target
Closes https://github.com/ProjectTech4DevAI/kaapi-backend/issues/935

* **New Features**
 Adds a standalone /guardrails endpoint that lets clients sanitise text via configured validators, independent of the LLM call flow.

  Product-facing

  - New POST /api/v1/guardrails — submit text + validator IDs, get back a job_id immediately.
  - New GET /api/v1/guardrails/{job_id} — poll for status and sanitised text.
  - Optional callback_url — webhook delivery of the sanitised result when the job completes.
  - request_metadata passthrough for client-side correlation.

## Checklist

Before submitting a pull request, please ensure that you mark these task.

- [ ] Ran `fastapi run --reload app/main.py` or `docker compose up` in the repository root and test.
- [ ] If you've fixed a bug or added code that is tested and has test cases.

## Summary by CodeRabbit

* **New Features**
  * Added a Guardrails API to submit text for validation and poll job status.
  * Support for optional asynchronous webhook delivery, including callback payload construction.
  * Extended job tracking with guardrails-specific metadata and a new guardrails job type.
* **Bug Fixes**
  * Improved safety-text handling when upstream responses omit sanitized output.
  * More consistent warning/error and usage shaping for callbacks and polling.
* **Documentation**
  * Added endpoint documentation for the Guardrails API and webhook flow.
* **Tests**
  * Added comprehensive route and service tests covering guardrails job lifecycle, callbacks, and edge cases.

### CODE REVIEW

File: backend/app/services/guardrails/jobs.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+                project_id=project_id,
+                organization_id=organization_id,
+            )
+            return {"success": False, "error": outcome.error}
+
+        safe_text = outcome.safe_text if outcome.safe_text is not None else request.text
+        # response_id is server-minted so callers always have a stable
+        # correlation handle, independent of whether the upstream guardrails
+        # service returns one.
+        response_id = str(uuid4())
```
Comment (Ayush8923) [inline backend/app/services/guardrails/jobs.py:364]: This `uuid4` is undefined. need to define above `from uuid import UUID, uuid4`

File: backend/app/services/guardrails/jobs.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
+import logging
+from typing import Any
+from uuid import UUID
```
Comment (Ayush8923) [inline backend/app/services/guardrails/jobs.py]:
```suggestion
from uuid import UUID, uuid4
```

File: backend/app/services/guardrails/jobs.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+            project_id=project_id,
+        )
+        span.set_attribute("guardrails.job_id", str(job.id))
+
+        # Persist the inbound request for traceability before the worker runs.
+        # We intentionally store the raw text: this endpoint exists to inspect
+        # potentially-unsafe content, so the request body itself is the audit log.
+        job_crud.update(
+            job_id=job.id,
+            job_update=JobUpdate(meta={"request": request.model_dump(mode="json")}),
```
Comment (Ayush8923) [inline backend/app/services/guardrails/jobs.py]: Right now, we are doing 2 commits here, like:
```
INSERT --> COMMIT

// Then again
UPDATE --> COMMIT
```

which is unnecessary because the row was just created.

Since `JobCrud.create()` now supports `meta`, pass it during creation itself:
```
job = job_crud.create(
    job_type=JobType.LLM_GUARDRAILS,
    trace_id=trace_id,
    project_id=project_id,
    meta={
        "request": request.model_dump(mode="json")
    },
)
```

Then we don't need this:
```
job_crud.update(
    job_id=job.id,
    job_update=JobUpdate(
        meta={"request": request.model_dump(mode="json")}
    ),
)
```


The final code looks like this:
```
job = job_crud.create(
    job_type=JobType.LLM_GUARDRAILS,
    trace_id=trace_id,
    project_id=project_id,
    meta={
        "request": request.model_dump(mode="json")
    },
)

span.set_attribute("guardrails.job_id", str(job.id))
```

File: backend/app/celery/tasks/job_execution.py [hunk 4]
```diff
@@ -94,6 +94,24 @@ def run_llm_chain_job(self, project_id: int, job_id: str, trace_id: str, **kwarg
     )
 
 
+@celery_app.task(bind=True, queue="high_priority", priority=9)
+@gevent_timeout(settings.CELERY_TASK_SOFT_TIME_LIMIT, "run_guardrails_job")
```
Comment (Ayush8923) [inline backend/app/celery/tasks/job_execution.py]: One thing: currently the traffic volume is expected to be maybe low from TAP, but I agree that guardrails jobs could eventually compete with interactive LLM requests. Let’s track this as a follow-up and evaluate a dedicated queue/lower priority once usage increases.

File: backend/app/services/guardrails/jobs.py [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+                error=str(e),
+                request_metadata=request.request_metadata,
+                project_id=project_id,
+                organization_id=organization_id,
+            )
+            return {"success": False, "error": str(e)}
+
+        warnings: list[str] = []
+        if outcome.bypassed:
+            warnings.append("guardrails_service_unavailable_text_returned_unchanged")
```
Comment (Ayush8923) [inline backend/app/services/guardrails/jobs.py]: For these types of warning crate the one constant: like:
```
GUARDRAILS_SERVICE_UNAVAILABLE = (
    "guardrails_service_unavailable_text_returned_unchanged"
)
```

and then use it here for better consistency.

File: backend/app/services/guardrails/jobs.py [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+        elif not outcome.raw and validators:
+            # apply_guardrails short-circuited with raw={} despite the caller
+            # sending non-empty validators — list_validators_config returned
+            # an empty list. The most common cause is the guardrails service
+            # being unreachable during the config-fetch step (it swallows
+            # transport errors and returns []). Surface this as a warning so
+            # callers do not mistake an effective no-op for a clean pass.
+            warnings.append("validator_configs_unresolved_text_returned_unchanged")
```
Comment (Ayush8923) [inline backend/app/services/guardrails/jobs.py]: Same for this, check this [comment](https://github.com/ProjectTech4DevAI/kaapi-backend/pull/939/changes#r3418812342).

File: backend/app/api/routes/guardrails.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+        callback_blob = meta.get("callback") if isinstance(meta, dict) else None
+
+        warnings: list[str] = []
+        if isinstance(callback_blob, dict):
+            raw_warnings = callback_blob.get("warnings")
+            if isinstance(raw_warnings, list):
+                warnings = [w for w in raw_warnings if isinstance(w, str)]
+
+        guardrails_response: GuardrailsCallbackData | None = None
+        if job.status.value == JobStatus.SUCCESS:
```
Comment (Ayush8923) [inline backend/app/api/routes/guardrails.py]: Can we do like this:
```
if job.status == JobStatus.SUCCESS:
```

File: backend/app/api/routes/guardrails.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+    dependencies=[
+        Depends(require_permission(Permission.REQUIRE_PROJECT)),
+        Depends(monitor_rate("llm_call")),
+    ],
+)
+def apply_guardrails_endpoint(
+    _current_user: AuthContextDep,
+    session: SessionDep,
+    request: GuardrailsRequest,
+):
```
Comment (Ayush8923) [inline backend/app/api/routes/guardrails.py]:
```suggestion
)  -> APIResponse[GuardrailsJobImmediatePublic]:
```

**Review by Ayush8923 — CHANGES_REQUESTED** [review]
Added few comments.

File: backend/app/alembic/versions/070_add_meta_and_guardrails_jobtype.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+
+def upgrade() -> None:
+    with op.get_context().autocommit_block():
+        op.execute("ALTER TYPE jobtype ADD VALUE IF NOT EXISTS 'LLM_GUARDRAILS'")
+
+    op.add_column(
+        "job",
+        sa.Column(
+            "meta",
```
Comment (vprashrex) [inline backend/app/alembic/versions/070_add_meta_and_guardrails_jobtype.py:26]: name it metadata not meta .. it is very confusing

File: backend/app/api/routes/guardrails.py [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+            )
+            value = safe_text if isinstance(safe_text, str) else (original_text or "")
+
+            response_id: str | None = None
+            if isinstance(callback_blob, dict):
+                rid = callback_blob.get("response_id")
+                if isinstance(rid, str):
+                    response_id = rid
+
+            guardrails_response = GuardrailsCallbackData.model_validate(
```
Comment (vprashrex) [inline backend/app/api/routes/guardrails.py:154]: I don't think so here model_validate would be necessary .. direct injection can also work .. since u are already doing parsing and validating it against its type

**Review by vprashrex — APPROVED** [review]

**Review by vprashrex — APPROVED** [review]
I couldn't find the `run_guardrails_job` function in `app.celery.tasks.job_execution`.

Could you please make sure that this function exists? I'm approving this from my side to avoid any blockers, but I wanted to double-check that the implementation is present.

**Comment by Prajna1999** [comment]
> I couldn't find the `run_guardrails_job` function in `app.celery.tasks.job_execution`.
> 
> Could you please make sure that this function exists? I'm approving this from my side to avoid any blockers, but I wanted to double-check that the implementation is present.

Its removed in the last merging with main. I will re add that

File: backend/app/services/guardrails/jobs.py [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+
+
+def start_job(
+    db: Session,
+    request: GuardrailsRequest,
+    project_id: int,
+    organization_id: int,
+) -> UUID:
+    """Create a guardrails-only job and schedule its Celery task."""
+    trace_id = correlation_id.get() or "N/A"
```
[the remark this answers is not in the document]
Comment (Ayush8923) [reply backend/app/services/guardrails/jobs.py:39]: @Prajna1999 I think we need to handle this.

File: backend/app/services/guardrails/jobs.py [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+                    job_id=job_uuid,
+                    job_update=JobUpdate(
+                        status=JobStatus.FAILED,
+                        error_message=outcome.error,
+                        meta={
+                            "request": request.model_dump(mode="json"),
+                            "response": outcome.raw,
+                        },
+                    ),
+                )
```
Comment (Ayush8923) [inline backend/app/services/guardrails/jobs.py]: Create the own function for this, because this is the hard block so it's easy to maintain and readability.

Comment (Ayush8923) [reply backend/app/services/guardrails/jobs.py]: Don't need to put everything in one function.

File: backend/app/services/guardrails/jobs.py [hunk 13]
```diff
@@ -0,0 +1,376 @@
...
+            f"callback_url={callback_url}, error={error}"
+        )
+        send_callback(
+            callback_url=callback_url,
+            data=payload,
+            webhook_secret=webhook_secret,
+        )
+
+
+def execute_job(
```
Comment (Ayush8923) [inline backend/app/services/guardrails/jobs.py:284]: this is too long function, can you please break this function in multiple sub small function so that this is easy to maintain and redability.

File: backend/app/services/guardrails/jobs.py [hunk 14]
```diff
@@ -0,0 +1,376 @@
...
+                cb_span.set_attribute("callback.status", "success")
+                logger.debug(
+                    f"[execute_job] Dispatching success callback | job_id={job_id}, "
+                    f"callback_url={callback_url}"
+                )
+                send_callback(
+                    callback_url=callback_url,
+                    data=callback_payload,
+                    webhook_secret=webhook_secret,
+                )
```
Comment (Ayush8923) [inline backend/app/services/guardrails/jobs.py]: Create the own subfunction for this, please check this [comment](https://github.com/ProjectTech4DevAI/kaapi-backend/pull/939/changes#r3481252689).

File: backend/app/services/guardrails/jobs.py (hunk 13 above)
Comment (Ayush8923) [reply backend/app/services/guardrails/jobs.py:284]: Refactor this function in a bit.

File: backend/app/services/llm/guardrails.py [hunk 15]
```diff
@@ -59,20 +181,45 @@ def run_guardrails_validation(
...
+    logger.info(
+        f"[run_guardrails_validation] POST guardrails | job_id={job_id}, url={url}, "
+        f"validators={len(validators)}, input_len={len(input_text)}, "
+        f"output_len={len(output_text) if output_text else 0}, "
+        f"payload_bytes={len(payload_bytes)}"
+    )
+    logger.info(
+        f"[run_guardrails_validation] Request payload | job_id={job_id}, "
+        f"payload={json.dumps(payload)}"
+    )
```
Comment (Ayush8923) [inline backend/app/services/llm/guardrails.py]: Don't add the multiple logger here, just create the one logger and add the important log info there.

File: backend/app/services/llm/guardrails.py [hunk 16]
```diff
@@ -59,20 +181,45 @@ def run_guardrails_validation(
...
+            elapsed_ms = int((time.monotonic() - started) * 1000)
+            logger.info(
+                f"[run_guardrails_validation] Response received | job_id={job_id}, "
+                f"status={response.status_code}, elapsed_ms={elapsed_ms}, "
+                f"response_bytes={len(response.content)}"
+            )
+            logger.info(
+                f"[run_guardrails_validation] Response body | job_id={job_id}, "
+                f"body={response.text}"
+            )
```
Comment (Ayush8923) [inline backend/app/services/llm/guardrails.py]: Update this too as per this [comment](https://github.com/ProjectTech4DevAI/kaapi-backend/pull/939/changes#r3481264273)

**Review by Ayush8923 — CHANGES_REQUESTED** [review]
Added the some more comment regarding refacoring.

File: backend/app/alembic/versions/070_add_meta_and_guardrails_jobtype.py (hunk 9 above)
Comment (Prajna1999) [reply backend/app/alembic/versions/070_add_meta_and_guardrails_jobtype.py:26]: since metadata is a default keyword for SQLModel, had to contend with meta

**Review by Ayush8923 — APPROVED** [review]
